### PR TITLE
Log stream responses with progress.

### PR DESCRIPTION
**Pull request by carlbraganza**
Fixes: https://github.com/kubernetes-csi/external-snapshot-metadata/issues/41

### CODE REVIEW

**Comment by carlbraganza** [comment]
Selective test output illustrating the logging:
```
=== RUN   TestStreamGetMetadataDeltaResponse/success
I1008 18:53:31.351617 1089848 util.go:362] GetSecretReference namespaceTemplate secret-ns, namespaceParams: map[volumesnapshot.namespace: volumesnapshotcontent.name:snapcontent-snap-2], resolved secret-ns
I1008 18:53:31.351628 1089848 util.go:391] GetSecretReference validated Secret: &SecretReference{Name:secret-name,Namespace:secret-ns,}
I1008 18:53:31.351642 1089848 get_metadata_delta.go:198] "stream response" op="GetMetadataDelta-1" namespace="" baseSnapshotName="snap-1" targetSnapshotName="snap-2" startingOffset=0 maxResults=2 blockMetadataType="FIXED_LENGTH" responseNum=1 volumeCapacityBytes=1073741824 blockMetadata="[{0,1024}]" numBlockMetadata=1
I1008 18:53:31.351649 1089848 get_metadata_delta.go:198] "stream response" op="GetMetadataDelta-1" namespace="" baseSnapshotName="snap-1" targetSnapshotName="snap-2" startingOffset=0 maxResults=2 blockMetadataType="FIXED_LENGTH" responseNum=2 volumeCapacityBytes=1073741824 blockMetadata="[{1,1024}{2,1024}]" numBlockMetadata=2
I1008 18:53:31.351654 1089848 get_metadata_delta.go:159] "stream EOF" op="GetMetadataDelta-1" namespace="" baseSnapshotName="snap-1" targetSnapshotName="snap-2" startingOffset=0 maxResults=2 blockMetadataType="FIXED_LENGTH" lastByteOffset=2 lastSize=1024 lastResponseNum=2 volumeCapacityBytes=1073741824
```
and
```
=== RUN   TestStreamGetMetadataAllocatedResponse/success
I1008 18:53:31.349395 1089848 util.go:362] GetSecretReference namespaceTemplate secret-ns, namespaceParams: map[volumesnapshot.namespace: volumesnapshotcontent.name:snapcontent-snap-1], resolved secret-ns
I1008 18:53:31.349419 1089848 util.go:391] GetSecretReference validated Secret: &SecretReference{Name:secret-name,Namespace:secret-ns,}
I1008 18:53:31.349448 1089848 get_metadata_allocated.go:177] "stream response" op="GetMetadataAllocated-1" namespace="" snapshotName="snap-1" startingOffset=0 maxResults=2 blockMetadataType="FIXED_LENGTH" responseNum=1 volumeCapacityBytes=1073741824 blockMetadata="[{0,1024}]" numBlockMetadata=1
I1008 18:53:31.349454 1089848 get_metadata_allocated.go:177] "stream response" op="GetMetadataAllocated-1" namespace="" snapshotName="snap-1" startingOffset=0 maxResults=2 blockMetadataType="FIXED_LENGTH" responseNum=2 volumeCapacityBytes=1073741824 blockMetadata="[{1,1024}{2,1024}]" numBlockMetadata=2
I1008 18:53:31.349459 1089848 get_metadata_allocated.go:138] "stream EOF" op="GetMetadataAllocated-1" namespace="" snapshotName="snap-1" startingOffset=0 maxResults=2 blockMetadataType="FIXED_LENGTH" lastByteOffset=2 lastSize=1024 lastResponseNum=2 volumeCapacityBytes=1073741824
```

File: pkg/internal/server/grpc/get_metadata_delta.go [hunk 1]
```diff
@@ -135,22 +137,76 @@ func (s *Server) convertToCSIGetMetadataDeltaRequest(ctx context.Context, req *a
...
+			lastSize = clientResp.BlockMetadata[numBlockMetadata].SizeBytes
+		}
+
+		if logger.V(HandlerTraceLogLevel).Enabled() {
+			var b strings.Builder
+			b.WriteString("[")
+			for _, bmd := range clientResp.BlockMetadata {
+				b.WriteString(fmt.Sprintf("{%d,%d}", bmd.ByteOffset, bmd.SizeBytes))
+			}
+			b.WriteString("]")
```
Comment (PrasadG193) [inline pkg/internal/server/grpc/get_metadata_delta.go:191]: Can we not use json encoding?

Comment (carlbraganza) [reply pkg/internal/server/grpc/get_metadata_delta.go:191]: JSON encoding is very verbose and would repeat the `byte_offset` and `size_bytes` labels for each entry. This is more compact and expresses the same data.

File: pkg/internal/server/grpc/get_metadata_allocated.go [hunk 2]
```diff
@@ -114,22 +116,76 @@ func (s *Server) convertToCSIGetMetadataAllocatedRequest(ctx context.Context, re
...
 		clientResp := s.convertToGetMetadataAllocatedResponse(csiResp)
+		blockMetadataType = clientResp.BlockMetadataType
+		volumeCapacityBytes = clientResp.VolumeCapacityBytes
+		numBlockMetadata = len(clientResp.BlockMetadata) - 1
+		if numBlockMetadata >= 0 {
+			lastByteOffset = clientResp.BlockMetadata[numBlockMetadata].ByteOffset
+			lastSize = clientResp.BlockMetadata[numBlockMetadata].SizeBytes
+		}
+
+		if logger.V(HandlerTraceLogLevel).Enabled() {
```
Comment (carlbraganza) [inline pkg/internal/server/grpc/get_metadata_allocated.go:164]: I am concerned whether we need another higher tracing level to display this data.

File: pkg/internal/server/grpc/server.go [hunk 3]
```diff
@@ -38,7 +38,8 @@ import (
 )
 
 const (
-	HandlerTraceLogLevel = 4
+	HandlerTraceLogLevel         = 4
+	HandlerDetailedTraceLogLevel = 5
```
Comment (carlbraganza) [inline pkg/internal/server/grpc/server.go:42]: I added a detailed trace level for the contents of the stream response loop.

Comment (PrasadG193) [reply pkg/internal/server/grpc/server.go:42]: @carlbraganza I think the logs level is not specific to handler methods, right? I mean the level can be set in log in any package. Should we give generic names like - debug, info, verbose, etc?

Comment (carlbraganza) [reply pkg/internal/server/grpc/server.go:42]: Klog uses a concept of a process wide verbosity level (the `V` function), and each -v flag increments the level used.   There are no clear definitions for info, debug, etc. - these are "severity" values in the log which could be logged at any verbosity level.

I defined these constants for the handler package, with the only constraint being that 
```
1 <= HandlerTraceLogLevel < HandlerDetailedTraceLogLevel
```
We could define common process wide constants to assign values to these constants later if we see the need, but for now I think most of the business logic is concentrated in the `grpc/server` package.

**Comment by carlbraganza** [comment]
I applied the permission change to `.prow.sh` and removed the invocation of the linter on the `test` target.

**Pull request merged**